src/System.Private.CoreLib/src/System/GC.cs

```diff
@@ -508,5 +512,57 @@ public static void EndNoGCRegion()
         {
             EndNoGCRegionWorker();
         }
+
+        // generation: -1 means to let GC decide (equivalent to regular new T[])
+        // 0 means to allocate in gen0
+        // GC.MaxGeneration means to allocate in the oldest generation
+        //
+        // pinned: true means you want to pin this object that you are allocating
+        // otherwise it's not pinned.
+        //
+        // alignment: only supported if pinned is true.
+        // -1 means you don't care which means it'll use the default alignment.
+        // otherwise specify a power of 2 value that's >= pointer size
+        // the beginning of the payload of the object (&array[0]) will be aligned with this alignment.
+        // static T[] AllocateArray<T>(int length, int generation = -1, bool pinned = false, int alignment = -1)
+        public static T[] AllocateArray<T>(int length, int generation = -1, bool pinned = false, int alignment = -1)
+        {
+            // TODO: VS resources
+            if (generation != -1)
+                throw new NotSupportedException();
+
+            if (pinned)
+                throw new NotSupportedException();
+
+            if (alignment != -1)
+                throw new NotSupportedException();
+
+            return AllocateArrayWorker<T>(length, clearMemory: true);
+        }
+
+        // Skips zero-initialization of the array if possible. If T contains object references, 
+        // the array is always zero-initialized.
+        public static T[] AllocateUninitializedArray<T>(int length, int generation = -1, bool pinned = false, int alignment = -1)
+        {
+            // TODO: VS resources
+            if (generation != -1)
+                throw new NotSupportedException();
+
+            if (pinned)
+                throw new NotSupportedException();
+
+            if (alignment != -1)
+                throw new NotSupportedException();
+
+            return AllocateArrayWorker<T>(length, clearMemory: false);
+        }
+
+        private static T[] AllocateArrayWorker<T>(int length, bool clearMemory)
+        {
+            if (length < 0)
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.lengths, 0, ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
+
+            return (T[])AllocateNewArray(typeof(T).TypeHandle.Value, length, clearMemory);
```

Comment:
This should pass in the typehandle for `T[]` so that we do not have to compute it hard way in the VM.

Reply:
Had that thought, but was not sure which way is harder. Will change.

---
In reply to: [277933188](https://github.com/dotnet/coreclr/pull/24096#discussion_r277933188) [](ancestors = 277933188)